graspologic/partition/leiden.py

```diff
@@ -0,0 +1,402 @@
+        networkx graph. This attribute corresponds to the edge data dict key.
+    is_weighted : Optional[bool]
+        Default is ``None``. Only used when creating a weighted edge list of tuples when the source graph is an
+        adjacency matrix. The graspologic.utils.to_weighted_edge_list() function will scan these matrices and attempt
+        to determine whether it is weighted or not. This flag can short circuit this test and the values in the
+        adjacency matrix will be treated as weights.
+    weight_default : float
+        Default is ``1.0``. If the graph is a networkx graph and the graph does not have a fully weighted sequence of
+        edges, this default will be used. If the adjacency matrix is found or specified to be unweighted, this
+        weight_default will be used for every edge.
+    check_directed : bool
+        Default is ``True``. If the graph is an adjacency matrix, we will attempt to ascertain whether it is directed
+        or undirected. As our leiden implementation is only known to work with an undirected graph, this function
+        will raise an error if it is found to be a directed graph. If you know it is undirected and wish to avoid this
+        scan, you can set this value to ``False`` and only the bottom left of the adjacency matrix will be used to
+        generate the weighted edge list.
+
+    Returns
+    -------
+    Dict[str, int]
+        The results of running leiden over the provided graph, a dictionary containing mappings of node -> community id.
+        The keys in the dictionary are the string representations of the nodes.
+
+    Raises
+    ------
+    ValueError
+    TypeError
+
+    See Also
+    --------
+    graspologic.utils.to_weighted_edge_list
+
+    References
+    ----------
+    .. [1] Traag, V.A.; Waltman, L.; Van, Eck N.J. "From Louvain to Leiden: guaranteeing well-connected communities",
+         Scientific Reports, Vol. 9, 2019
+    .. [2] https://github.com/microsoft/graspologic-native
+
+    Notes
+    -----
+    This function is implemented in the `graspologic-native` Python module, a module written in Rust for Python.
+
+    """
+    _validate_common_arguments(
+        starting_communities,
+        iterations,
+        resolution,
+        randomness,
+        use_modularity,
+        random_seed,
+        is_weighted,
+        weight_default,
+        check_directed,
+    )
+    graph = _validate_and_build_edge_list(
+        graph, is_weighted, weight_attribute, check_directed, weight_default
+    )
+
+    if iterations < 1:
+        raise ValueError("iterations must be > 1")
+
+    _improved, _modularity, partitions = gn.leiden(
+        edges=graph,
+        starting_communities=starting_communities,
+        resolution=resolution,
+        randomness=randomness,
+        iterations=iterations,
+        use_modularity=use_modularity,
+        seed=random_seed,
+    )
+
+    return partitions
+
+
+class HierarchicalCluster(NamedTuple):
```

Comment:
not a fan of this name, there are many hierarchical cluster algos... e.g. https://scikit-learn.org/stable/modules/generated/sklearn.cluster.AgglomerativeClustering.html as the most common example which most people call hierarchical clustering.

another comment is to what extent can we have a unified interface with `DivisiveCluster`

Reply:
Describing a HierarchicalCluster has nothing to do with the algorithm though.  It's just an entry of "node to cluster, and cluster to previous cluster if any".  That we use it in hierarchical leiden shouldn't be too much fo a problem, and if anyone else wants to use this in another algorithm, they certainly can.

The long and the short of this class is that it is a capture of the node->cluster mapping at the moment it is created, like a log entry of sorts.  As it transitions from one cluster to another, we keep track of where it came from previously, but there's nothing tying it to hierarchical leiden per se (except that hierarchical leiden happens to return a list of 'em)

Regardless, the name shouldn't be seen as Leiden only, and if it makes sense to emit the same thing from DivisiveClustering, that's great!  We just make use of both the final clustering dictionary AND the history of all the cluster transitions as we run through hierarchical leiden ( https://github.com/microsoft/graspologic-native/blob/dev/packages/network_partitions/src/leiden/hierarchical.rs#L120 )) for different sorts of visualizations (showing different partitioning based on level is very useful at drilling through clusters, we've found)